src/store.cc

```diff
@@ -540,7 +540,17 @@ storeGetPublic(const char *uri, const HttpRequestMethod& method)
 StoreEntry *
 storeGetPublicByRequestMethod(HttpRequest * req, const HttpRequestMethod& method, const KeyScope keyScope)
 {
-    return Store::Root().find(storeKeyPublicByRequestMethod(req, method, keyScope));
+    StoreEntry *e = Store::Root().find(storeKeyPublicByRequestMethod(req, method, keyScope));
+    if (e) {
+        const mayCollapse = EBIT_TEST(e->flags, ENTRY_FWD_HDR_WAIT) ||
+                            Store::Root().smpAware() && (e->hasTransients() && (!e->hasMemStore() && !e->hasDisk()));
```

Comment:
We should figure out here how to distinguish public entries created in 'master' collapsed transactions. It is easy for non-SMP case (just check for ENTRY_FWD_HDR_WAIT). However, I have not found a simple way how to it for SMP case.

Reply:
I think you are on the right track with these conditions. Here are a few adjustments:

1. s/mayCollapse/foundCfInitiator/
2. Remove smpAware() if it is not needed.
3. If smpAware() is needed, please fix parenthesis to surround the whole second sub-condition.

BTW, how come we did not have this problem when setting crSlave? In other words, how does processExpired() know that the result of its storeGetPublicByRequest() call is the CF-initiating entry and not just a regular Store entry?

Reply:
I think we do not have this problem in `processExpired()` because of `ksRevalidation` key scope. An entry with this scope exist only until the revalidation response is received.